### ISSUE TITLE
Elasticsearch-Cluster gets automatically reinstalled and loses all PersistentVolumes

### ISSUE THREAD

**Issue by davelosert** [elastic/cloud-on-k8s#3032]
#### Our Setup
We have the `elastic-operator` installed in Version 1.0.1 on our Azure AKS Cluster and use it for a 3-replica elasticsearch cluster.

The operator is within the `elastic-system` namespace while our cluster operaters in the `prod-elasticsearch` namespace.

#### What happened?
Yesterday,  for the second time in two months, the whole elasticsearch-cluster was suddenly and automatically reinstalled out of nowhere - meaning all Pods, Secrets, Services, PVC and PVs got deleted and  recreated. 

Upon recreation, the cluster also entered a failed state as the third replica of the statefulset was never created.

We can not explain why this happened. None of us was working during the time. We don't have any other services that could have done this. Our Audit Logs show no manual intervention of any kind - so it has to have been triggered from wtihin the kubernetes-cluster.

#### Obseravations from Logging
The operator logs unfortunately don't give us much insight (see below). From the Logs, it was doing it's normal reconciliation runs until it suddenly started creating secrets followed by the rest of the resources as if the elasticsearch cluster was never there. There are also no `events` whatsoever listed in the operator or the `elasticsearch`-Resource. 

I attached the [operator-logs.log](https://github.com/elastic/cloud-on-k8s/files/4609091/controller-logs.log) from the timeframe it happened to this issue. The `statefulset` never gets healthy as the third pod actually never gets created seemingly due to some 503 to the license server.

Unfortunately, we don't have any logs from the pre-deletion cluster as we have no centralized logging yet, so we are kind of blind here.

We could only fix the failing state by completely deleting the cluster and reinstall it manually again.

#### Impact
Since all `PVs` were gone, all our data was lost. Luckily, we are still only in a staging environment, so that was not business critical. 

However, we'd still like to know why this happened - so I am opening this issue to maybe get some help (and maybe report a severe bug if it happenes to be the operator doing this 😉 )

**Do you have any idea how this can happen? Have you seen such behavior before?**

**Environment**

* ECK version: 1.0.1

* Kubernetes information: AKS Version 1.17.3

```
$ kubectl version
Client Version: version.Info{Major:"1", Minor:"15", GitVersion:"v1.15.5", GitCommit:"20c265fef0741dd71a66480e35bd69f18351daea", GitTreeState:"clean", BuildDate:"2019-10-15T19:16:51Z", GoVersion:"go1.12.10", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"17", GitVersion:"v1.17.3", GitCommit:"260aa4feed9c12e5256e5c69c2b329d768de2f26", GitTreeState:"clean", BuildDate:"2020-04-03T15:19:08Z", GoVersion:"go1.13.6", Compiler:"gc", Platform:"linux/amd64"}
```
* Elasticserach Configuration:
```
---
apiVersion: elasticsearch.k8s.elastic.co/v1
kind: Elasticsearch
metadata:
  name: prod-elasticsearch
  namespace: prod-elasticsearch
spec:
  version: 7.6.0
  http:
    tls:
      selfSignedCertificate:
        subjectAltNames:
          # Add Full Cluster DNS entry to SAN so the certificate can be used from other namespaces as well
          - dns: prod-elasticsearch-es-http.prod-elasticsearch.svc.cluster.local
  nodeSets:
  - name: prod-elasticsearch
    config:
      node.master: true
      node.data: true
      node.ingest: true
      node.ml: true
      node.store.allow_mmap: false
    podTemplate:
      metadata:
        labels:
          app: elasticsearch
          app_env: prod
      spec:
        nodeSelector:
          storagetier: Premium_LRS
        containers:
        - name: elasticsearch
          resources:
            limits:
              memory: 1Gi
            requests:
              cpu: 250m
              memory: 1Gi
          env:
          - name: ES_JAVA_OPTS
            value: -Xms500m -Xmx500m
    count: 3
    volumeClaimTemplates:
    - metadata:
        name: elasticsearch-data
      spec:
        accessModes:
        - ReadWriteOnce
        resources:
          requests:
            storage: 17Gi
        storageClassName: premium-expandable # Storage of type azure-disk
[controller-logs.log](https://github.com/elastic/cloud-on-k8s/files/4609088/controller-logs.log)

```

* Resource definition: https://download.elastic.co/downloads/eck/1.0.1/all-in-one.yaml

* Logs:

```
2020-05-10T08:23:47.556909366Z {"level":"info","@timestamp":"2020-05-10T08:23:47.548Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":527,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                 │
│2020-05-10T08:24:01.373467132Z {"level":"info","@timestamp":"2020-05-10T08:24:01.373Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":527,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":13.824174529}                                                                               │
│2020-05-10T08:24:01.373637333Z {"level":"error","@timestamp":"2020-05-10T08:24:01.373Z","logger":"controller-runtime.controller","message":"Reconciler error","ver":"1.0.1-bcb74688","controller":"elasticsearch-controller","request":"prod-elasticsearch/prod-elasticsearch","error":"etcdserver: request timed out","errorCauses":[{"error":"etcdserver: request time│
│2020-05-10T08:24:02.373997052Z {"level":"info","@timestamp":"2020-05-10T08:24:02.373Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":528,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                 │
│2020-05-10T08:24:03.082517043Z {"level":"info","@timestamp":"2020-05-10T08:24:03.082Z","logger":"zen2","message":"Ensuring no voting exclusions are set","ver":"1.0.1-bcb74688","namespace":"prod-elasticsearch","es_name":"prod-elasticsearch"}                                                                                                                        │
│2020-05-10T08:24:03.15794227Z {"level":"info","@timestamp":"2020-05-10T08:24:03.157Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":528,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.783964419}                                                                                 │
│2020-05-10T09:18:08.091158005Z {"level":"info","@timestamp":"2020-05-10T09:18:08.090Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":633,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                       │
│2020-05-10T09:18:08.097425954Z {"level":"info","@timestamp":"2020-05-10T09:18:08.097Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":633,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.006533551}                                                                                      │
│2020-05-10T10:18:08.097824702Z {"level":"info","@timestamp":"2020-05-10T10:18:08.097Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":634,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                       │
│2020-05-10T10:18:08.107438277Z {"level":"info","@timestamp":"2020-05-10T10:18:08.107Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":634,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.009697576}                                                                                      │
│2020-05-10T11:18:08.107952832Z {"level":"info","@timestamp":"2020-05-10T11:18:08.107Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":635,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                       │
│2020-05-10T11:18:08.124374759Z {"level":"info","@timestamp":"2020-05-10T11:18:08.124Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":635,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.016583128}                                                                                      │
│2020-05-10T12:18:08.125016844Z {"level":"info","@timestamp":"2020-05-10T12:18:08.124Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":636,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                       │
│2020-05-10T12:18:08.132304701Z {"level":"info","@timestamp":"2020-05-10T12:18:08.132Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":636,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.007283357}                                                                                      │
│2020-05-10T12:27:02.633545985Z {"level":"info","@timestamp":"2020-05-10T12:27:02.633Z","logger":"webhook-certificates-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":218,"namespace":"","name":"elastic-webhook.k8s.elastic.co"}                                                                                                │
│2020-05-10T12:27:02.684210978Z {"level":"info","@timestamp":"2020-05-10T12:27:02.684Z","logger":"webhook-certificates-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":218,"namespace":"","name":"elastic-webhook.k8s.elastic.co","took":0.050905495}                                                                               │
│2020-05-10T12:27:02.752936811Z {"level":"info","@timestamp":"2020-05-10T12:27:02.740Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":529,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                 │
│2020-05-10T12:27:02.752982111Z {"level":"info","@timestamp":"2020-05-10T12:27:02.740Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":637,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                       │
│2020-05-10T12:27:02.752989012Z {"level":"info","@timestamp":"2020-05-10T12:27:02.745Z","logger":"license-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":637,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.00520044}                                                                                       │
│2020-05-10T12:27:03.162338686Z {"level":"info","@timestamp":"2020-05-10T12:27:03.162Z","logger":"webhook-certificates-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":219,"namespace":"elastic-system","name":"elastic-webhook-server-cert"}                                                                                     │
│2020-05-10T12:27:03.178767713Z {"level":"info","@timestamp":"2020-05-10T12:27:03.178Z","logger":"webhook-certificates-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":219,"namespace":"elastic-system","name":"elastic-webhook-server-cert","took":0.016479328}                                                                    │
│2020-05-10T12:27:03.708576721Z {"level":"info","@timestamp":"2020-05-10T12:27:03.708Z","logger":"zen2","message":"Ensuring no voting exclusions are set","ver":"1.0.1-bcb74688","namespace":"prod-elasticsearch","es_name":"prod-elasticsearch"}                                                                                                                        │
│2020-05-10T12:27:03.805927276Z {"level":"info","@timestamp":"2020-05-10T12:27:03.805Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":529,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":1.06523966}                                                                                 │
│2020-05-10T12:27:03.805985876Z {"level":"info","@timestamp":"2020-05-10T12:27:03.805Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":530,"namespace":"prod-nas-service","name":"elasticsearch-ca-cert"}                                                                                                │
│2020-05-10T12:27:03.806294779Z {"level":"info","@timestamp":"2020-05-10T12:27:03.806Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":530,"namespace":"prod-nas-service","name":"elasticsearch-ca-cert","took":0.000460803}                                                                               │
│2020-05-10T12:27:03.806311179Z {"level":"info","@timestamp":"2020-05-10T12:27:03.806Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":531,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                 │
│2020-05-10T12:27:04.670882183Z {"level":"info","@timestamp":"2020-05-10T12:27:04.670Z","logger":"zen2","message":"Ensuring no voting exclusions are set","ver":"1.0.1-bcb74688","namespace":"prod-elasticsearch","es_name":"prod-elasticsearch"}                                                                                                                        │
│2020-05-10T12:27:04.760270076Z {"level":"info","@timestamp":"2020-05-10T12:27:04.760Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":531,"namespace":"prod-elasticsearch","name":"prod-elasticsearch","took":0.953915196}                                                                                │
│2020-05-10T12:27:11.741585508Z {"level":"info","@timestamp":"2020-05-10T12:27:11.741Z","logger":"elasticsearch-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":532,"namespace":"prod-elasticsearch","name":"prod-elasticsearch"}                                                                                                 │
│2020-05-10T12:27:11.747857556Z {"level":"info","@timestamp":"2020-05-10T12:27:11.747Z","logger":"generic-reconciler","message":"Creating resource","ver":"1.0.1-bcb74688","kind":"Secret","namespace":"prod-elasticsearch","name":"prod-elasticsearch-es-elastic-user"}                                                                                                 │
│2020-05-10T12:27:12.228995687Z {"level":"info","@timestamp":"2020-05-10T12:27:12.228Z","logger":"generic-reconciler","message":"Updating resource","ver":"1.0.1-bcb74688","kind":"Secret","namespace":"prod-elasticsearch","name":"prod-elasticsearch-es-xpack-file-realm"}                                                                                             │
│2020-05-10T12:27:12.263757357Z {"level":"info","@timestamp":"2020-05-10T12:27:12.263Z","logger":"generic-reconciler","message":"Creating resource","ver":"1.0.1-bcb74688","kind":"Service","namespace":"prod-elasticsearch","name":"prod-elasticsearch-es-prod-elasticsearch"}
```

**Comment by charith-elastic** [comment]
That is very strange indeed. The following log line at 08:24:01 is interesting:
```
{"level":"error","@timestamp":"2020-05-10T08:24:01.373Z","logger":"controller-runtime.controller","message":"Reconciler error","ver":"1.0.1-bcb74688","controller":"elasticsearch-controller","request":"
prod-elasticsearch/prod-elasticsearch","error":"etcdserver: request timed out"}
```

This suggests that the etcd server was either under heavy load or was down completely. As the state is stored in etcd, I wonder whether there was some state corruption that resulted from it. For almost an hour since this line was logged (08:24 to 09:18), there's no activity at all recorded in the logs -- which is highly unusual and suspicious. It looks like your Kubernetes cluster underwent something that halted the operator for an hour and possibly corrupted the stored state.

```
2020-05-10T08:24:03.15794227Z {"level":"info","@timestamp":"2020-05-10T08:24:03.157Z","logger":"elasticsearch-controller","message":"Ending reconciliation run","ver":"1.0.1-bcb74688","iteration":528,"namespace":"prod-elasticsearch",
"name":"prod-elasticsearch","took":0.783964419}
2020-05-10T09:18:08.091158005Z {"level":"info","@timestamp":"2020-05-10T09:18:08.090Z","logger":"license-controller","message":"Starting reconciliation run","ver":"1.0.1-bcb74688","iteration":633,"namespace":"prod-elasticsearch","na
me":"prod-elasticsearch"}
```

Do you have access to the API server logs and any other logs or metrics that could shed a light on the state of the Kubernetes control plane during this period? 

It would help to know more details about your cluster such as the number of nodes, number of resources (pods, secrets, configmaps etc.), and CPU, memory and disk utilization as well.

**Comment by anyasabo** [comment]
>Our Audit Logs show no manual intervention of any kind - so it has to have been triggered from wtihin the kubernetes-cluster.

To be clear, do you mean the API server audit logs? If so, they should show what user deleted the resources.

>The statefulset never gets healthy as the third pod actually never gets created seemingly due to some 503 to the license server.

This can be checked by doing a describe on the pod -- there will likely be events to see why it is not starting successfully. ECK does not create the pod directly, it creates a statefulset that creates the pod.

It also may be worth checking to see if you have say, copied secrets (say, the elastic user credentials) between namespaces or otherwise have resources with owner references that span namespaces. This can trigger a [bug](https://github.com/kubernetes/kubernetes/issues/65200) in the Kubernetes garbage collector with similar behavior to what you described.

**Comment by davelosert** [comment]
First of all: Thank you both very much for your fast responses!

I am going to answer @anyasabo first because well: 
We actually DO copy a secret, namely the`elasticsearch-es-http-certs-public`, in order to be able to access elasticsearch with TLS internally from other namespaces with the self-signed certificate.

We basically use this command to do it:

```bash
kubectl get secret prod-elasticsearch-es-http-certs-public --namespace prod-elasticsearch -o yaml \
		| sed s/\"namespace: prod-elasticsearch\"/\"namespace: internal-services\"/\
		| kubectl apply -n internal-services -f -"
```

I just checked and yes - the ownerReferences are similar. That might just be it. I will definitly try to change hits.

> To be clear, do you mean the API server audit logs? If so, they should show what user deleted the resources.

I actually meant the Azure's Audit Logs. There all we can see is that it was the cluster itself deleting the storage volumes.

> This can be checked by doing a describe on the pod -- there will likely be events to see why it is not starting successfully. ECK does not create the pod directly, it creates a statefulset that creates the pod.

Thanks for the hint. Since I already reinstalled the cluster, I can not check it anymore, but I will remember next time!


@charith-elastic 
Also very interesting about etcd - I overlooked this.
I checked on some other services, but none seem to have been disrupted in any case. Happy and normal logging during the timeframe (i checked 08:20 - 12:30).

> Do you have access to the API server logs and any other logs or metrics that could shed a light on the state of the Kubernetes control plane during this period?

Unfortunatley no, we are currently collecting no metrics at all and do not have any API- or Master Logs enabled. But this just went up on our priority list to implement rather sooner than later ;) 

> It would help to know more details about your cluster such as the number of nodes, number of resources (pods, secrets, configmaps etc.), and CPU, memory and disk utilization as well.

This is what I can give you without metrics collection:
- Number of Nodes: 5, each with 2 CPUs and 16 GB RAM, 260 GB Storage
- Number of Pods / ConfigMaps / Secrets : 64 / 26 / 141
- Number of PVs: 15
- Memory Utilization: 64 % on average accross all nodes (almost evenly distributed)
- CPU Utilization: 68% on average accorrs all nodes (almost evenly distributed)


## To sum it up
Thanks again very much as this was already very helpful. 👏 

Right now, I'd suspect that the secret-copying might actually be the culprit. Some other people in the issue reported similar problems. So this will be the first thing I'll fix.

Besides this, I guess with our missing metrics and logging it will probably hard to go any deeper than this. ☹️  Unless you have any other hinch or idea as to the cause, I'd suggest that we will improve metrics and logging in our cluster and - in case this happens again - I'll provide better and more infos here.

**Comment by anyasabo** [comment]
The method you used to copy secrets will trigger the bug pretty reliably, so it seems likely that is the cause. Removing the owner reference as part of the copy step should prevent it going forward.

The way to confirm would be looking at the API server audit logs (you would see the garbage collector deleting the underlying elasticsearch resources), but it appears they need to be explicitly enabled in AKS: https://github.com/Azure/AKS/issues/946

**Comment by davelosert** [comment]
Thanks again for the feedback.

And yes - in Azure this is the way to go. We actually already had it enabled and pushed it to Azure Log Analytics. But the pricing for this is horrific, so we disabled it until we have our own logging stroage enabled (probably elasticsearch ;) )

**Comment by charith-elastic** [comment]
Looks like this was most likely caused by the upstream bug in Kubernetes. Please re-open if there are any updates that suggest otherwise.
